Implement GPUBuffer.mapAsync and update wgpu-core

kunalmohan opened the pull request:
<!-- Please describe your changes on the following line: -->
This PR covers the following-
1. update wgpu-core to use serializable `RenderPass` and `ComputePass`. Implement `DispatchIndirect`.
2. Implement `GPUBuffer.mapAsync()`.

The `mapAsync()` methods has the following flow-
1. A new promise is created in `mapAsync()` and a message sent to the server-side. This returns a promise.
2. On the server-side, a `struct BufferMapInfo` object stores the relevant info required to send data to the client-side on completion of mapping. This object is stored in a `buffer_maps: HashMap<id::BufferId, BufferMapInfo<'a>>,` in `WGPU`. `buffers_maps` basically stores the data for buffers with pending mapping.
3. `buffer_map_async()` command is issued with a callback responsible to send the data back to the client.
4. All the devices are polled every time the server receives a new message.
5. The data is sent back on completion and promise resolved through `AsyncWGPUListener` (similar to how the `requestDevice()` and `RequestAdapter` are processed).
6. On receiving the data, the client sends back a `BufferMapComplete` message to the server to remove the entry from `buffer_maps` hashmap.

I intended to implement `getMappedRange()`, too, in this PR but it's quite a task in itself (implementing intersecting ArrayBuffers). This already being a 500+ line PR, I thought it best to get this model approved first (also it will probably be easier to review this and `getMappedRange` separately). `getMappedRange` would come up in the next one.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

kunalmohan committed:
Implement GPUBuffer.mapAsync and update wgpu-core

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webidls/GPUComputePassEncoder.webidl, components/script/dom/gpucomputepassencoder.rs, components/script/dom/gpumapmode.rs, components/script/dom/gpubuffer.rs, components/script/dom/webidls/GPUBuffer.webidl and 10 more
 * @kichjang: components/script/dom/webidls/GPUComputePassEncoder.webidl, components/script/dom/gpucomputepassencoder.rs, components/script/dom/gpumapmode.rs, components/script/dom/gpubuffer.rs, components/script/dom/webidls/GPUBuffer.webidl and 10 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

kunalmohan commented:
r?@kvark

kunalmohan committed:
Remove peek-poke and peek-poke-derive from ignored packages

kvark commented on components/script/dom/gpudevice.rs:
```diff
@@ -284,28 +227,16 @@ impl GPUDeviceMethods for GPUDevice {
-                        } else {
-                            Some(NonZeroU64::new(bind.minBufferBindingSize).unwrap())
-                        },
+                        min_binding_size: wgt::BufferSize::new(bind.minBufferBindingSize),
```
oh this is niiice!

kvark commented on components/script/dom/gpudevice.rs:
```diff
@@ -433,11 +364,7 @@ impl GPUDeviceMethods for GPUDevice {
-                                } else {
-                                    wgt::BufferSize::WHOLE
-                                },
+                                size: wgt::BufferSize::new(b.size.unwrap_or(0)),
```
nit: super bikesheddy, it could be `b.size.and_then(wgt::BufferSize::new)`

kvark commented on components/webgpu/lib.rs:
```diff
@@ -296,20 +314,69 @@ impl WGPU {
+            self.devices.iter().for_each(|device| {
+                let global = &self.global;
+                let device_id = device.0;
+                gfx_select!(device_id => global.device_poll(device_id, false));
```
would be nice to move it out and poll less often

kvark commented on components/webgpu/lib.rs:
```diff
@@ -296,20 +314,69 @@ impl WGPU {
-    fn run(mut self) {
+    fn run(&'a mut self) {
         while let Ok(msg) = self.receiver.recv() {
+            self.devices.iter().for_each(|device| {
```
we can use [poll_all_devices](https://github.com/gfx-rs/wgpu/blob/29e286bdc1474457c2a34155d559e30268df74ad/wgpu-core/src/device/mod.rs#L2733) instead

kvark commented on components/webgpu/lib.rs:
```diff
@@ -296,20 +314,69 @@ impl WGPU {
+                    self.buffer_maps.insert(buffer_id, map_info);
+
+                    unsafe extern "C" fn callback(status: BufferMapAsyncStatus, userdata: *mut u8) {
+                        let info = (userdata as *mut BufferMapInfo).as_ref().unwrap();
```
This looks totally unsafe, the lifetime of `a` can't be respected if the stuff is unsafely casted from C.
Maybe just turn it into a `NonNull` pointer?

kvark commented on components/webgpu/lib.rs:
```diff
@@ -296,20 +314,69 @@ impl WGPU {
+                                    warn!("Could not send BufferMapAsync Response ({})", e);
+                                }
+                            },
+                            _ => warn!("Could not map buffer({:?})", info.buffer_id),
```
we should `error!` in these cases, since we aren't really handling them gracefully

kvark commented on components/webgpu/lib.rs:
```diff
@@ -296,20 +314,69 @@ impl WGPU {
+                                let global = info.global;
+                                let data_pt = gfx_select!(info.buffer_id =>
+                                    global.buffer_get_mapped_range(info.buffer_id, 0, None));
+                                let data = slice::from_raw_parts(data_pt, info.size).to_vec();
```
are we sending an immutable slice? That would only work for read mapping, not write mappings.

kunalmohan replied:
I am a bit confused about what your concern here is. We are sending a vector to the client. If I have to change the mutability I'll have to do it on the client-side while receiving, i.e. `WebGPUResponse::BufferMapAsync(mut bytes)` in `handle_response()`? And we'll be writing back any data in the `WebGPURequest::Unmap` via `write_buffer()` or 'set_buffer_sub_data()`. How does that change things here?

kvark replied:
Here is how I see this working, at this point at least:
  1. for a write-mappable buffer, the implementation carries around the shared memory with its data. When user asks for mapping, you make the server to notify the client when the GPU is done, and you just expose this shmem to the user to modify, followed by `set_buffer_sub_data` on the server side.
  2. for a read-mappable buffer, the server waits for GPU, copies the immutable slice out into a shmem that is given to the user.

I suppose the code we see here is for (2) only? In this case, why do we have `host_map` parameter to the message?

kunalmohan replied:
I see. I had a different approach in my mind. I was going with sending vectors across the processes for data exchange between client and server.
I guess you are pointing towards using [`IpcSharedMemory`](https://docs.rs/ipc-channel/0.14.1/ipc_channel/ipc/struct.IpcSharedMemory.html) (similar to the previous `MapReadAsync` implementation).

kunalmohan replied:
Sorry. I see there is a bigger concern here than the `IpcSharedMemory`. But it would probably be solved when I figure out the intersecting ArrayBuffers part.

kvark reviewed (changes requested):
Very nice! Just a few notes here

kunalmohan commented on components/script/dom/gpubuffer.rs:
```diff
@@ -79,14 +88,21 @@ impl GPUBuffer {
         state: GPUBufferState,
-        size: GPUSize64,
-        usage: u32,
+        size: u64,
```
Nothing in particular. I had removed the size field initially while cleaning up the extra validation for `copyBuffertoBuffer`. Then, while implementing `mapAsync` I had to reintroduce it so I just used u64 instead of importing GPUSize64. I'll change it if it looks more appropriate.

kunalmohan commented on components/script/dom/gpubuffer.rs:
```diff
@@ -202,3 +269,40 @@ impl GPUBufferMethods for GPUBuffer {
+    fn handle_response(&self, response: WebGPUResponse, promise: &Rc<Promise>) {
+        match response {
+            WebGPUResponse::BufferMapAsync(bytes) => {
+                if unsafe {
```
The return type for `ArrayBuffer::create()` is `Result<(),()>` so there is no actual error to log here. Logging a custom message could still be helpful so I'll do that.

kvark replied:
ah ok. In cases like this `match` is still better (it's better in most cases), since you can match on `Err(())` to explicitly show that the error is `()` and avoid people asking weird questions :)

kunalmohan committed:
address review comments

kvark commented on components/script/dom/gpubuffer.rs:
```diff
@@ -189,7 +207,56 @@ impl GPUBufferMethods for GPUBuffer {
+        let map_range = if size == 0 {
+            offset..self.size
+        } else {
+            offset..size
```
please resolve the things when you push the code that fixes them, not earlier
otherwise, it's hard to know when to verify resolution

kunalmohan replied:
Sorry :sweat_smile: . I'll keep that in mind.

kunalmohan commented:
I have addressed most of the review comments. I am still unsure about the last one- about how that would be done. I'll try to dig into it but maybe it will be better to solve that along with `get_mapped_range()` depending how the data will be exposed to the user. I can convert this to a draft and continue implementing `get_mapped_range()` here itself if that's a better idea.

kvark reviewed (approved):
We had a chat about this on "#wgpu:matrix.org", and it looks like the PR is on the right track:)

kunalmohan committed:
Ensure proper unmap of buffer

kunalmohan commented:
@kvark I have changed the Unmapping process to use `get_mapped_range()` to write back the data to buffer while unmapping in case of `mapAsync` as well as `mapped_at_creation`.

kvark reviewed (approved):
Looks reasonable, thank you!

kvark commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 183a43d has been approved by `kvark`

<!-- @bors-servo r=kvark 183a43daa9ae0378c3c347bad74398f9121c3a03 -->

kunalmohan commented:
```
error[E0599]: no method named `handle_mut` found for struct `dom::bindings::trace::RootedTraceableBox<mozjs_sys::jsgc::Heap<*mut js::jsapi::JSObject>>` in the current scope
    --> components/script/dom/gpubuffer.rs:284:62
     |
284  |                         MutableHandle::from_raw(self.mapping.handle_mut()),
     |                                                              ^^^^^^^^^^ method not found in `dom::bindings::trace::RootedTraceableBox<mozjs_sys::jsgc::Heap<*mut js::jsapi::JSObject>>`
     | 
    ::: components/script/dom/bindings/trace.rs:1036:1
     |
1036 | pub struct RootedTraceableBox<T: 'static + JSTraceable> {
     | ------------------------------------------------------- method `handle_mut` not found for this
error[E0599]: no method named `handle_mut` found for struct `dom::bindings::trace::RootedTraceableBox<mozjs_sys::jsgc::Heap<_>>` in the current scope
    --> components/script/dom/gpudevice.rs:189:53
     |
189  |                     MutableHandle::from_raw(mapping.handle_mut()),
     |                                                     ^^^^^^^^^^ method not found in `dom::bindings::trace::RootedTraceableBox<mozjs_sys::jsgc::Heap<_>>`
     | 
    ::: components/script/dom/bindings/trace.rs:1036:1
     |
1036 | pub struct RootedTraceableBox<T: 'static + JSTraceable> {
     | ------------------------------------------------------- method `handle_mut` not found for this
```
checks failed. I didn't get this on my system earlier, but on a rebase against master I can see it failing now. I'll rectify this and push the changes.

kunalmohan committed:
Fix ArrayBuffer creation in buffer mapping

kvark commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit db2d313 has been approved by `kvark`

<!-- @bors-servo r=kvark db2d313a1b76886c656a90280bfe4f44b909d977 -->

bors-servo commented:
:hourglass: Testing commit db2d313a1b76886c656a90280bfe4f44b909d977 with merge 1ea332e41a0ddc777c44004654fd62eb7828be27...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/X8MR62eZSpiIilEmDOoD8g)

jdm commented:
@bors-servo retry
* #25522

bors-servo commented:
:hourglass: Testing commit db2d313a1b76886c656a90280bfe4f44b909d977 with merge af110ac21fcf1e108c919f5e9d724b2441996ed6...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/CeGTFmdBS8mu3-Mfgdpf8w)
Approved by: kvark
Pushing af110ac21fcf1e108c919f5e9d724b2441996ed6 to master...
